### ISSUE TITLE
Program stuck when using an async/task operation before an `EntryPoint` instruction

### ISSUE THREAD

**Issue by MangelMaxime** [dotnet/fsharp#15627]
Please provide a succinct description of the issue.

**Repro steps**

Provide the steps required to reproduce the problem:

1. Create an `Build.fsproj` file with the following content

	```xml
	<Project Sdk="Microsoft.NET.Sdk">
	  <PropertyGroup>
	    <OutputType>Exe</OutputType>
	    <TargetFramework>net6.0</TargetFramework>
	  </PropertyGroup>
	  <ItemGroup>
	    <Compile Include="Build.fs" />
	  </ItemGroup>
	</Project>
	```

2. Create a `Build.fs` file with the following content

	```fs
	open System.IO
	
	let deployPath = Path.GetFullPath "deploy"
	
	printfn "1"
	
	async {
	     printfn "2 %s" deployPath
	}
	|> Async.RunSynchronously
	
	[<EntryPoint>]
	let main args =
	    printfn "3"
	    0
	```

3. Run `dotnet run` and see that the program print `1` and then get stuck.

Here is a zip, with the everything already set up. You can unzip it and run `dotnet run`

[Reproduction.zip](https://github.com/dotnet/fsharp/files/12092846/Reproduction.zip)

**Expected behavior**

We should see:

```
1
2 XXXX
3
```

**Actual behavior**

Program get stuck after printing `1`

**Known workarounds**

Move the `async` instruction inside the `EntryPoint`

Remove the `EntryPoint` instruction.

**Related information**

Provide any related information (optional):

* Operating system: macOS 13.4.1 (c) (22F770820d)
* .NET Runtime kind (.NET Core, .NET Framework, Mono) .NET Core 6 or 8 preview
* Editing Tools (e.g. Visual Studio Version, Visual Studio)

**Comment by vzarytovskii** [comment]
Everything top level, which needs "running" (i.e. not bindings) ends up in startup code class constructor and gets blocked, and never exits the static constructor.

**Comment by MangelMaxime** [comment]
Indeed, moving the `let deployPath` declaration inside the async solve the problem.

However, in my real use case, I am using the project as a build system and kind of need the variable declaration outside, in order to avoid code duplication.

It is also worth noting that I am not sure if this is a "real" bug or not. I would say yes because the code compile but doesn't run as expected but I can be wrong.
